### PR TITLE
Fix nested pressable's event propagation.

### PR DESCRIPTION
**Pull request by latekvo**
### Issue:
`Test case`: `3` nested pressables
`Actions`: `click`, `press for <500ms`, `press for >500ms`
`Expected behaviour`: No propagation
`Observed behaviour`: Propagation for `500ms` (until long press activates), then no propagation.

### Cause:
Touch and non-active events disregard component's layer,
only state respecting component's level is `ACTIVE`.

### Solution:
Create a component which activates immediately after pressing `Pressable`, capture that event and use it for state setting and callback calling.

### Important:
This PR also performs a lot of de-cluttering work in the `utils.ts` file.
This is necessary due to the convoluted nature of the handled problem, it was especially important to me in this PR to be dealing with clearly labelled functions.

closes: #2980

### CODE REVIEW

**Comment by latekvo** [comment]
As of [7326626](https://github.com/software-mansion/react-native-gesture-handler/pull/2981/commits/73266264ae18e977e5859d32094f041fc36c4c04), both Android and IOS works thanks to View's `onTouchStart` which is identical to Gesture handler's `onTouchStart` except for `stopPropagation()` method which prevents events from reaching obstructed parent elements.

**Comment by latekvo** [comment]
Previously I described how `InnerPressableEvent` (formerly `PressEvent`) has a structure that looks like it should be recursive but isn't.

Namely:
```js
type InnerPressableEvent = {
  changedTouches: InnerPressableEvent[];
  touches: InnerPressableEvent[];
  [...]
};
```

After further testing, `View` actually uses it's version of this type in a recursive manner, which isn't the case for the legacy `Pressable`.

Since recursiveness in this case is not forced - aligns with the type's structure, and does occur on other components using this structure I'd argue that this is a direct upgrade from the current state, and thus should be implemented onto `Pressable` despite lack of it's presence on the legacy component.

**Comment by latekvo** [comment]
As of [1b245fa](https://github.com/software-mansion/react-native-gesture-handler/pull/2981/commits/1b245facd3265ac84427475962285ad9ad3d4968), android, ios and web are all working.

**Comment by latekvo** [comment]
Please excuse the review request, i've found more issues in some of the older examples.

**Comment by latekvo** [comment]
I fixed all the bugs i've found so far, this PR is up for review once again. @m-bert @j-piasecki 
My only concern now is a substantially delayed response on the web, but it does not impede on the rest of the functionality, and will be either resolved or explained later on. 

For now I only know it's directly caused by the gesture simply calling `onTouchesUp` 1 second later than it should, but I still have to figure out what causes this behaviour.

File: src/components/Pressable/utils.ts [hunk 1]
```diff
@@ -146,7 +192,7 @@ const innerStyleKeys = new Set([
   'paddingVertical',
   'start',
   'end',
-  'direction', // iOS only
+  'direction', // IOS only
```
Comment (m-bert) [inline src/components/Pressable/utils.ts:195]: This should be left as `iOS` 😅 

```suggestion
  'direction', // iOS only
```

File: src/components/Pressable/Pressable.tsx [hunk 2]
```diff
@@ -52,68 +51,77 @@ export default function Pressable(props: PressableProps) {
...
-          props.onHoverOut?.(adaptStateChangeEvent(event));
+          props.onHoverOut?.(gestureToPressableEvent(event));
         }),
     [props]
   );
 
   const pressDelayTimeoutRef = useRef<number | null>(null);
+  const propagationGreenLight = useRef<boolean>(false);
+
+  // IOS only, propagationGreenLight setting occurs after other checks when in scroll-view
```
Comment (m-bert) [inline src/components/Pressable/Pressable.tsx:93]:
```suggestion
  // iOS only, propagationGreenLight setting occurs after other checks when in scroll-view
```

**Comment by latekvo** [comment]
On `iOS` the small delay between click and activation in the `Nested Pressables` example is caused by using native elements inside a scrollable parent. It is normal and completely unavoidable for all iOS native components.

**Comment by latekvo** [comment]
Currently `LongPress` is used for gathering `touch events` in `Pressable`.
This is done because unlike `Manual` gesture (which theoretically would be a better fit), it works consistently across all platforms.

My question is, is it better to keep using `LongPress` despite not using it's state handler, or better to change the code to work with `Manual` gesture, which would have no unnecessary overhead inside caused by state handlers?

I think using `LongPress` is overall better - it makes the code a little bit shorter and likely prevents some unforeseen bugs, as it doesn't require as much platform specific code.
On the other hand, `Manual` gesture looks cleaner, and is potentially more optimal.

(c.c. @m-bert @j-piasecki )

File: src/components/Pressable/Pressable.tsx [hunk 3]
```diff
@@ -183,42 +243,90 @@ export default function Pressable(props: PressableProps) {
...
+                // that means we are in a no-scroll, no-nesting situation
+                // and native start may be trated as a replacement to the measure function.
+                propagationGreenLight.current = true;
+                isBoundsCheckNecessary.current = false;
+              }
+            }
+          }
+          if (Platform.OS === 'web') {
+            propagationGreenLight.current = true;
+          }
```
Comment (m-bert) [inline src/components/Pressable/Pressable.tsx:327]: We have to do something with this `if-else` hell 😅

Below you can find my suggestion, though I may have missed something.

```suggestion
          if (Platform.OS === 'web') {
            propagationGreenLight.current = true;
          }

          if (Platform.OS !== 'ios') {
            return;
          }

          if (awaitingEventPayload.current) {
            propagationGreenLight.current = true;

            if (hasPassedBoundsChecks.current) {
              pressInHandler(awaitingEventPayload.current);
              awaitingEventPayload.current = null;
            } else {
              pressOutHandler(awaitingEventPayload.current);
              propagationGreenLight.current = false;
            }

            return;
          }

          if (hasPassedBoundsChecks.current) {
            propagationGreenLight.current = true;
            return;
          }

          if (preventNativeEffects.current) {
            preventNativeEffects.current = false;
            return;
          }

          // In case Native Start is the very first gesture on iOS to launch
          // that means we are in a no-scroll, no-nesting situation
          // and native start may be trated as a replacement to the measure function.
          propagationGreenLight.current = true;
          isBoundsCheckNecessary.current = false;
```
```suggestion
          // iOS sets ACTIVE state on press down
          if (Platform.OS === 'ios') {
            if (awaitingEventPayload.current) {
              propagationGreenLight.current = true;
              if (hasPassedBoundsChecks.current) {
                pressInHandler(awaitingEventPayload.current);
                awaitingEventPayload.current = null;
              } else {
                pressOutHandler(awaitingEventPayload.current);
                propagationGreenLight.current = false;
              }
            } else {
              if (hasPassedBoundsChecks.current) {
                propagationGreenLight.current = true;
              } else {
                if (preventNativeEffects.current) {
                  preventNativeEffects.current = false;
                  return;
                }

                // In case Native Start is the very first gesture on iOS to launch
                // that means we are in a no-scroll, no-nesting situation
                // and native start may be trated as a replacement to the measure function.
                propagationGreenLight.current = true;
                isBoundsCheckNecessary.current = false;
              }
            }
          }
          if (Platform.OS === 'web') {
            propagationGreenLight.current = true;
          }
```

File: src/components/Pressable/Pressable.tsx [hunk 4]
```diff
@@ -183,42 +243,90 @@ export default function Pressable(props: PressableProps) {
...
           if (handlingOnTouchesDown.current) {
-            onEndHandlingTouchesDown.current = () => pressOutHandler(event);
+            onEndHandlingTouchesDown.current = () =>
+              pressOutHandler(gestureTouchToPressableEvent(event));
             return;
           }
-
-          pressOutHandler(event);
+          // On iOS, when pressed out before Native gesture pressed in,
+          // without setting this variable Native gesture would think it's the first one to be launched
```
Comment (m-bert) [inline src/components/Pressable/Pressable.tsx:251]: Is there a way to somehow rephrase this sentence? It is not the easiest one to read 😅

**Review by m-bert — COMMENTED** [review]
Overall looks good, we can think of making `nestedPressables` example look a bit better 😅

File: src/components/Pressable/utils.ts [hunk 5]
```diff
@@ -38,16 +43,22 @@ const touchToPressEvent = (
...
-  absoluteY: event.absoluteY,
+  >,
+  timestamp: number,
+  targetId: number
+): InnerPressableEvent => ({
+  identifier: event.handlerTag,
+  locationX: event.x,
+  locationY: event.y,
+  pageX: event.absoluteX,
+  pageY: event.absoluteY,
```
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/utils.ts:57]: Thanks for the review,

`InnerPressableEvent` allows us to return events compatible with the legacy `Pressable` events while maintaining consistency between different platforms, all of which returned a slightly different version of the `GestureResponderEvent` on the original `Pressable`. 

`InnerPressableEvent` is our version of `NativeTouchEvent`, which occupies the `nativeElement` field of the `GestureResponderEvent`.

It was nearly impossible to recreate the entire `GestureResponderEvent` due to the large number of fields, multiple conflicting types, functions which couldn't be replicated and it's deeply nested structure, when constructing those events from regular `GestureHandler` events, I don't remember why I had to rewrite the `NativeTouchEvent` into `InnerPressableEvent`. 

I think at first it was to add missing fields from the  `NativeTouchEvent` but currently I see no difference between the two, and I think we can actually replace all usages of `InnerPressableEvent` with just  `NativeTouchEvent`.

File: src/components/Pressable/Pressable.tsx [hunk 6]
```diff
@@ -52,56 +54,64 @@ export default function Pressable(props: PressableProps) {
...
+
   const pressInHandler = useCallback(
-    (event: GestureTouchEvent) => {
-      props.onPressIn?.(adaptTouchEvent(event));
+    (event: PressableEvent) => {
+      if (handlingOnTouchesDown.current) {
+        awaitingEventPayload.current = event;
+      }
+
+      if (propagationGreenLight.current === false) {
```
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:106]: It's done for readability, I think `=== false` is much more readable, especially with refs which can be mistakenly used like this:

```js
if (propagationGreenLight)
```

In which case they are always truthy, and the linter doesn't tell you anything is wrong.

File: src/components/Pressable/Pressable.tsx [hunk 7]
```diff
@@ -52,56 +54,64 @@ export default function Pressable(props: PressableProps) {
...
+          props.onHoverOut?.(gestureToPressableEvent(event));
         }),
     [props]
   );
 
   const pressDelayTimeoutRef = useRef<number | null>(null);
+  const propagationGreenLight = useRef<boolean>(false);
+
+  // iOS only, propagationGreenLight setting occurs after other checks when in scroll-view
+  const awaitingEventPayload = useRef<PressableEvent | null>(null);
```
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:98]: I agree the word itself implies a boolean, maybe `awaited` would work better?

This `ref` holds an asynchronously awaited payload.
It's not certain when or if this event payload will be set, and I think `awaited` really highlights that behaviour.

File: src/components/Pressable/Pressable.tsx [hunk 8]
```diff
@@ -52,56 +54,64 @@ export default function Pressable(props: PressableProps) {
...
-          props.onHoverOut?.(adaptStateChangeEvent(event));
+          props.onHoverOut?.(gestureToPressableEvent(event));
         }),
     [props]
   );
 
   const pressDelayTimeoutRef = useRef<number | null>(null);
+  const propagationGreenLight = useRef<boolean>(false);
+
+  // iOS only, propagationGreenLight setting occurs after other checks when in scroll-view
```
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:97]: No, it's intended for `awaitingEventPayload`.

But I see it is very poorly worded which makes it look like it's misplaced,
maybe a comment like this would work better?

```suggestion
  // iOS only, due to variable flow of gestures on iOS, their events sometimes have to be saved for later use
```

File: src/components/Pressable/Pressable.tsx [hunk 9]
```diff
@@ -34,7 +34,9 @@ export default function Pressable(props: PressableProps) {
 
   // Disabled when onLongPress has been called
   const isPressCallbackEnabled = useRef<boolean>(true);
-  const isPressedDown = useRef<boolean>(false);
+  const hasPassedBoundsChecks = useRef<boolean>(false);
+  const isBoundsCheckNecessary = useRef<boolean>(true);
```
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:38]: I will have to do more testing, 

I remember it was being used to remove the need for `onTouchesDown` when `Native` gesture's `onStart` ran first, and I'm not sure why it's not being used anymore.

File: src/components/Pressable/Pressable.tsx (hunk 4 above)
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:251]: Oh yeah I see the issue 😄 
Do you think this would work?
```suggestion
          // On iOS, short taps will make `LongPress` gesture call `onTouchesUp` before `Native` gesture calls  `onStart` 
          // This variable makes sure that `onStart` is not detected as the first gesture since the `Pressable` was pressed.
```

File: src/components/Pressable/Pressable.tsx (hunk 3 above)
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:327]: I agree, I played around with flattening all of these `ifs` and ended up with something similar to the first suggestion, so I think the first suggestion makes the most sense.

File: src/components/Pressable/Pressable.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:38]: Ok, there is absolutely no need for that flag, as the case it would be optimising is already instant.
I'll remove it.

File: src/components/Pressable/Pressable.tsx [hunk 10]
```diff
@@ -52,56 +54,64 @@ export default function Pressable(props: PressableProps) {
...
             return;
           }
-          props.onHoverOut?.(adaptStateChangeEvent(event));
+          props.onHoverOut?.(gestureToPressableEvent(event));
         }),
     [props]
   );
 
   const pressDelayTimeoutRef = useRef<number | null>(null);
+  const propagationGreenLight = useRef<boolean>(false);
```
[the remark this answers is not in the document]
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx:95]: Second one sounds better 👍

File: src/components/Pressable/Pressable.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx:97]: What about
```suggestion
  // For iOS only: due to the varying flow of gestures, events sometimes have to be saved for later use
```

or simply:

```suggestion
  // iOS only: due to the varying flow of gestures, events sometimes have to be saved for later use```

File: src/components/Pressable/Pressable.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx:98]: Makes sense 👍

File: src/components/Pressable/Pressable.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx:106]: IMHO it doesn't improve readability and I'd stick to using just `!` prefix (cc @j-piasecki)

File: src/components/Pressable/Pressable.tsx [hunk 11]
```diff
@@ -52,56 +54,64 @@ export default function Pressable(props: PressableProps) {
...
+    (event: PressableEvent) => {
+      if (handlingOnTouchesDown.current) {
+        awaitingEventPayload.current = event;
+      }
+
+      if (propagationGreenLight.current === false) {
+        return;
+      }
+
+      // If iOS passes the propagationGreenLight by here,
```
[the remark this answers is not in the document]
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx]: I don't think we have to remove it, we could rephrase it to sound a bit more natural. WDYT @j-piasecki?

File: src/components/Pressable/Pressable.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:96]: I like the last one the most 👍

File: src/components/Pressable/Pressable.tsx (hunk 4 above)
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx:247]: Yeah, that's better. I'd change _**"makes sure"**_ to _**"ensures"**_ though

File: src/components/Pressable/Pressable.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:97]: Hey so would `waitingEventPayload` work fine? wdyt?

[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx]: I ended up choosing `waitingEventPayload` as (besides `eventPayload`) it's the only non-ambiguous choice out of the ones brought up. 
- `cachedEventPayload` conveyed different message
- `savedEventPayload` had the same issue of sounding like a `boolean` value

[the remark this answers is not in the document]
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx]: WDYT @j-piasecki?

File: src/components/Pressable/utils.ts [hunk 12]
```diff
@@ -119,6 +130,41 @@ const adaptTouchEvent = (event: GestureTouchEvent): PressableEvent => {
...
+    changedTouches: changedTouchesList,
+  };
+};
+
+const nativeTouchToPressableEvent = (
+  event: GestureResponderEvent
+): PressableEvent => ({
+  nativeEvent: nativeTouchToInnerPressableEvent(event.nativeEvent),
+});
+
```
Comment (j-piasecki) [inline src/components/Pressable/utils.ts:167]: This seems unused

File: src/components/Pressable/Pressable.tsx [hunk 13]
```diff
@@ -142,9 +162,36 @@ export default function Pressable(props: PressableProps) {
...
+  const longPressMinDuration =
+    (props.delayLongPress ?? DEFAULT_LONG_PRESS_DURATION) +
+    (props.unstable_pressDelay ?? 0);
+
+  const pressAndTouchGesture = useMemo(
     () =>
-      Gesture.Manual()
+      Gesture.LongPress()
+        .minDuration(Number.MAX_SAFE_INTEGER) // Stops long press from blocking native gesture
+        .maxDistance(Number.MAX_SAFE_INTEGER) // Stops long press from cancelling after set distance
```
Comment (j-piasecki) [inline src/components/Pressable/Pressable.tsx:193]: Why do we even need LongPress in this case? Since the actual callback is invoked by a timeout anyway, cannot it be handled by the native gesture?

File: src/components/Pressable/utils.ts (hunk 12 above)
Comment (latekvo) [reply src/components/Pressable/utils.ts]: You're right, I'm removing it.

File: src/components/Pressable/Pressable.tsx (hunk 13 above)
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:193]: Native gesture has an unpredictable behaviour, which requires large amounts of additional code for edge case checking to synchronise between the different platforms, and this also affects touches - I checked. 
On the other hand, the core gestures as far as I see, behave identically between all the platforms, which is why i chose to use `LongPress`.

Of course, if you believe there is a good reason to really try and get rid of it in favour of either `Manual` gesture, or `Native` gesture, I can do that. I just think it may add anywhere from 50 to 150 new lines of code, and may cause some unforeseen bugs caused by newly added edge cases.

For example, when debugging the `Native` gesture on iOS, I composed a list of edge cases all of which had to be individually found through rigorous testing, and had to be uniquely addressed:
 
- inside scrollable
  - nested
    - click for <=1frame (before measure)
    - click for >1frame <200ms (before native)
    - click for >200ms <500ms (before long press)
    - click for >500ms (with long press)
  - not-nested
    - click for <=1frame (before measure)
    - click for >1frame <200ms (before native)
    - click for >200ms <500ms (before long press)
    - click for >500ms (with long press)
- not in scrollable
  - nested
    - click for >200ms <500ms (before long press)
    - click for >500ms (with long press)
  - not-nested
    - click for >1frame <200ms (before native)
    - click for >200ms <500ms (before long press)
    - click for >500ms (with long press)

I referenced this choice further in [this comment](https://github.com/software-mansion/react-native-gesture-handler/pull/2981#issuecomment-2233477313)

File: src/components/Pressable/Pressable.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/components/Pressable/Pressable.tsx]: I don't see a problem with `waitingEventPayload, but I can add one more to the list of candidates - `deferredEventPayload`. Though, no strong opinions as well.

[the remark this answers is not in the document]
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx]: IMHO `deferred` sound ok 👍

[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx]: I agree, changing it to `deferred`

File: src/components/Pressable/Pressable.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx]: note - `propagationGreenLight` has been renamed to `isTouchPropagationAllowed` since

File: src/components/Pressable/Pressable.tsx (hunk 13 above)
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx:193]: During some further testing with @j-piasecki we found issues with changing `LongPress` to `Manual` or moving the logic to `Native`, and found that not much logic can be moved either way.
And so we decided to leave the code as it is for now, thus marking this as resolved.

File: src/components/Pressable/Pressable.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx]: Okay, I've  just discussed it with @j-piasecki and we agreed that we should stay consistent. In other places in this file we have for example:

```tsx
if (hoverInTimeout.current) {
  ...
}
```

so I believe we can use prefix `!` instead of comparing it to `false`, if it is just for readability sake 😅

[the remark this answers is not in the document]
Comment (latekvo) [reply src/components/Pressable/Pressable.tsx]: fixed in [64e9d50](https://github.com/software-mansion/react-native-gesture-handler/pull/2981/commits/64e9d50462b800b9787563b3ed9bceff4fa61bc8)

File: src/components/Pressable/Pressable.tsx [hunk 14]
```diff
@@ -34,7 +34,8 @@ export default function Pressable(props: PressableProps) {
 
   // Disabled when onLongPress has been called
   const isPressCallbackEnabled = useRef<boolean>(true);
-  const isPressedDown = useRef<boolean>(false);
+  const hasPassedBoundsChecks = useRef<boolean>(false);
+  const preventNativeEffects = useRef<boolean>(false);
```
Comment (m-bert) [inline src/components/Pressable/Pressable.tsx:38]:
```suggestion
  const shouldPreventNativeEffects = useRef<boolean>(false);
```

**Review by m-bert — APPROVED** [review]
Finally, the time has come :shipit:

File: src/components/Pressable/Pressable.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (m-bert) [reply src/components/Pressable/Pressable.tsx]: Now when I look at it it was removed as we don't use `propagationGreenLight` name anymore

**Review by j-piasecki — APPROVED** [review]
